webrtc/README.md

```diff
@@ -0,0 +1,439 @@
+these messages, consult the [QUIC RFC].
+
+``` proto
+syntax = "proto2";
+
+package webrtc.pb;
+
+message Message {
+  enum Flag {
+    // The sender will no longer send messages on the stream.
+    FIN = 0;
+    // The sender will no longer read messages on the stream. Incoming data is
+    // being discarded on receipt.
+    STOP_SENDING = 1;
+    // The sender abruptly terminates the sending part of the stream. The
+    // receiver can discard any data that it already received on that stream.
+    RESET_STREAM = 2;
+  }
+
+  optional Flag flag=1;
+
+  optional bytes message = 2;
+}
+```
+
+Note that in contrast to QUIC (see [QUIC RFC - 3.5 Solicited State
+Transitions](https://www.rfc-editor.org/rfc/rfc9000.html#section-3.5)) a libp2p
+WebRTC endpoint receiving a `STOP_SENDING` frame SHOULD NOT send a
+`RESET_STREAM` frame in reply. The `STOP_SENDING` frame is used for accurate
+accounting of the number of bytes sent for connection-level flow control in
+QUIC. The libp2p WebRTC message framing is not concerned with flow-control and
+thus does not need the `RESET_STREAM` frame to be send in reply to a
+`STOP_SENDING` frame.
+
+Encoded messages including their length prefix MUST NOT exceed 16kiB to support
+all major browsers. See ["Understanding message size
+limits"](https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API/Using_data_channels#understanding_message_size_limits).
+Implementations MAY choose to send smaller messages, e.g. to reduce delays
+sending _flagged_ messages.
+
+### Ordering
+
+Implementations MAY expose an unordered byte stream abstraction to the user by
+overriding the default value of `ordered` `true` to `false` when creating a new
+data channel via
+[`RTCPeerConnection.createDataChannel`](https://www.w3.org/TR/webrtc/#dom-peerconnection-createdatachannel).
+
+### Open Questions
+
+- There are two ways to open a `RTCDataChannel`, either via `negotiated: true`
+  and thus out-of-band channel ID negotiation or `negotiated: false` and thus
+  in-band channel ID negotiation.
+
+  Is the assumption correct that in-band stream ID negotiation requires an
+  additional round trip? If so how about using out-of-band channel id
+  negotiation only (thus no additional round trip) based on the DTLS role?
+
+  > When one side wants to open a channel using out-of-band negotiation, it
+  > picks a stream. Unless otherwise defined or negotiated, the streams are
+  > picked based on the DTLS role (the client picks even stream identifiers, and
+  > the server picks odd stream identifiers).
+
+  https://www.rfc-editor.org/rfc/rfc8831#name-opening-a-data-channel
+
+- Do browsers allow reuse of `RTCDataChannel` IDs? If so, should we allow libp2p
+  WebRTC implementations to reuse channel IDs of previously closed channels?
+  Motivation would be to have more than 2^16 channels throughout the lifetime of
+  a WebRTC connection. Note that this would still restrict a connection to at
+  most 2^16 concurrent channels.
```

Comment:
@ckousik @lightsofapollo or anyone else: Do you know whether major browsers support reuse of `RTCDataChannel` IDs? If not, could you validate this with Chrome and Firefox?

_Sparked by @marten-seemann comment in https://github.com/libp2p/specs/pull/412#issuecomment-1255472209._

Reply:
https://source.chromium.org/chromium/chromium/src/+/main:third_party/webrtc/pc/data_channel_controller.cc;l=336;drc=0d0cbbeedf5831fe32e4823c0e5a854983c64ed8 libwebrtc releases ids once datachannels are closed, so they can be reused.

Reply:
This is now documented with 3315b5c. Let me know what you think.

Reply:
I am resolving here. Note that there is still an open question by @melekes in the sibling discussion. See https://github.com/libp2p/specs/pull/412#discussion_r984746570.